File: src/data_rover/core/table/schema.py
"""The `kind='table'` artifact payload schema.

A table's rows are TUPLES OF BINDINGS: the row source contributes one binding
slot (scope / navigation) or N (chains, one per chain column), and every
`expand` column contributes one more. Each column names a `source` — an earlier
binding slot or an earlier column — that resolves to an ordered set of elements;
the column maps over it. `collapse` keeps the mapped values in one cell;
`expand` promotes them to a new binding slot (one row per value).

Static validation here rejects cycles (a ColumnRef must point strictly
backward), non-element navigation sources, multi-binding element sources, and
chain_index on a non-chains row source. Two further rules need the metamodel or
the resolved navigation and are checked at evaluation time (see evaluate.py):
expand-on-a-scalar-property and chain_index-out-of-range.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, Field, TypeAdapter, model_validator

from data_rover.core.navigation.schema import NavigationDefinition
from data_rover.core.script.schema import SnippetSource
from data_rover.core.search.criteria import Criterion

SCHEMA_VERSION = 1
MAX_COLUMNS = 50


class NavigationSource(BaseModel):
    """At most one of `ref` (saved artifact id) / `definition` (inline).

    NEITHER set (`{}`) is a legal, UNCONFIGURED source: the table editor
    creates a navigation column / row source before the user picks its
    navigation, and rejecting that transient state here would 422 every
    evaluate/save of the WHOLE table until it is configured. Evaluation
    treats an unconfigured source as reaching nothing (empty cell / no rows)
    — same tolerant-evaluation stance as `cells.expand_property_values`.
    BOTH set stays rejected: the request is ambiguous, not incomplete.
    """

    ref: str | None = None
    definition: NavigationDefinition | None = None

    @model_validator(mode="after")
    def _at_most_one(self) -> NavigationSource:
        if self.ref is not None and self.definition is not None:
            raise ValueError("provide at most one of `ref` / `definition`")
        return self

    @property
    def is_empty(self) -> bool:
        """True for the unconfigured (`{}`) source."""
        return self.ref is None and self.definition is None


# ---- row source -------------------------------------------------------------
class ScopeRows(BaseModel):
    kind: Literal["scope"] = "scope"
    types: list[str] = Field(default_factory=list)
    criteria: list[Criterion] = Field(default_factory=list)


class NavigationRows(BaseModel):
    kind: Literal["navigation"] = "navigation"
    navigation: NavigationSource
    step_index: int | None = None


class ChainRows(BaseModel):
    kind: Literal["chains"] = "chains"
    navigation: NavigationSource


RowSource = Annotated[
    ScopeRows | NavigationRows | ChainRows, Field(discriminator="kind")
]


# ---- column source ----------------------------------------------------------
class RowSlot(BaseModel):
    kind: Literal["row"] = "row"
    chain_index: int = Field(default=0, ge=0)


class ColumnRef(BaseModel):
    kind: Literal["column"] = "column"
    index: int = Field(ge=0)
    #: Only legal when the referenced column is a NAVIGATION column: resolve
    #: this reference to the elements at THIS chain step (0 = the chain's
    #: start) instead of the referenced column's own projected step. None =
    #: the referenced column's own behavior (unchanged).
    step_index: int | None = None


ColumnSource = Annotated[RowSlot | ColumnRef, Field(discriminator="kind")]


# ---- columns ----------------------------------------------------------------
class JsonColumnOptions(BaseModel):
    """Per-column JSON-export settings (spec:
    docs/superpowers/specs/2026-07-25-table-json-export-design.md).

    Lives on the COLUMN rather than on `TableDefinition` as an index-keyed map
    deliberately: column indices move under reorder/insert/remove (the frontend
    already carries `remapTableSortForRemove/Move/Insert` to keep a single index
    valid across those edits), and settings attached to the column travel with
    it for free.

    The field is named `json_export` and not `json`: pydantic v2 still carries a
    deprecated `.json()` method that a field of that name would collide with.
    """

    #: "" means "derive from the header" — see `resolve_json_keys`.
    key: str = ""
    #: Key for the column's OWN value inside the array entries `group`
    #: produces. Meaningful only where `group` is honored; "" falls back to
    #: the RESOLVED group key — which is exactly what the single key this
    #: field splits already did, so old definitions export unchanged.
    item_key: str = ""
    #: How an element reference renders. Ignored by columns that never produce
    #: elements (a property column), which is tolerated rather than rejected.
    value: Literal["name", "id", "object"] = "name"
    #: Roll this `expand` column's rows back up into one array. Honored only on
    #: a VISIBLE EXPAND column; ignored elsewhere (the column editor can flip
    #: expand->collapse at any time and a 422 would block the whole export).
    group: bool = False


class ColumnExportOptions(BaseModel):
    """Per-column export overrides (spec:
    docs/superpowers/specs/2026-07-28-table-export-settings-design.md).

    On the COLUMN rather than in an index-keyed map on `TableDefinition`, for
    the same reason `JsonColumnOptions` is: column indices move under reorder,
    insert, and remove, and settings attached to the column travel with it for
    free.

    Presentation-only, like `hidden` and `json_export`: never consulted during
    evaluation.
    """

    #: `None` = follow `hidden`, which is what makes every pre-existing
    #: definition export byte-identically. `True` on a hidden column exports it
    #: WITHOUT unhiding it in the grid; `False` on a visible one keeps it on
    #: screen and out of the file.
    include: bool | None = None
    #: xlsx header override; "" keeps today's `header or kind`. JSON renames
    #: through `json_export.key` instead and ignores this — one rename box per
    #: format, never two on one row.
    header: str = ""


class RowNumberExportOptions(BaseModel):
    """Export overrides for the row-number pseudo-column.

    Lives on the definition rather than on a column because there is no
    `Column` to hang it off: the row-number column is synthesized by the
    renderers from `show_row_numbers`.
    """

    include: bool = True
    #: "" -> "#" (xlsx).
    header: str = ""
    #: "" -> "row_number" (JSON).
    key: str = ""


class JsonSplitOptions(BaseModel):
    """Split the JSON export into one file per base element (spec:
    docs/superpowers/specs/2026-08-13-table-export-split-and-custom-export-design.md,
    docs/superpowers/specs/2026-08-19-custom-export-v2-design.md).

    Presentation-only, like `export_order`: never consulted during evaluation.
    `filename_template` must contain `${name}` — enforced at EXPORT time
    (`core/table/split.validate_template`), deliberately not here, so a stored
    bad template cannot block saving or evaluating the table."""

    enabled: bool = False
    #: `${name}` is replaced by each base element's display name.
    filename_template: str = ""


class ElementColumn(BaseModel):
    kind: Literal["element"] = "element"
    source: ColumnSource = Field(default_factory=RowSlot)
    header: str = ""
    width_px: int | None = None
    #: Presentation-only: a hidden column is still evaluated (later columns
    #: may reference it via ColumnRef) but is omitted from the grid and the
    #: xlsx export. Never feed this into evaluation — dropping the column
    #: would shift ColumnRef indices and the expand-slot arithmetic.
    hidden: bool = False
    #: JSON-export settings; `None` means "all defaults", which keeps saved
    #: payloads clean for the overwhelming majority of columns.
    json_export: JsonColumnOptions | None = None
    #: Export overrides (inclusion, xlsx header). `None` means "all defaults",
    #: which keeps saved payloads clean for the overwhelming majority.
    export: ColumnExportOptions | None = None


class PropertyColumn(BaseModel):
    kind: Literal["property"] = "property"
    source: ColumnSource = Field(default_factory=RowSlot)
    name: str
    mode: Literal["collapse", "expand"] = "collapse"
    keep_empty: bool = True
    header: str = ""
    width_px: int | None = None
    #: Presentation-only: a hidden column is still evaluated (later columns
    #: may reference it via ColumnRef) but is omitted from the grid and the
    #: xlsx export. Never feed this into evaluation — dropping the column
    #: would shift ColumnRef indices and the expand-slot arithmetic.
    hidden: bool = False
    #: JSON-export settings; `None` means "all defaults", which keeps saved
    #: payloads clean for the overwhelming majority of columns.
    json_export: JsonColumnOptions | None = None
    #: Export overrides (inclusion, xlsx header). `None` means "all defaults",
    #: which keeps saved payloads clean for the overwhelming majority.
    export: ColumnExportOptions | None = None


class NavigationColumn(BaseModel):
    kind: Literal["navigation"] = "navigation"
    source: ColumnSource = Field(default_factory=RowSlot)
    navigation: NavigationSource
    step_index: int | None = None
    mode: Literal["collapse", "expand"] = "collapse"
    keep_empty: bool = True
    sort_mode: Literal["value", "count"] = "value"
    cell_cap: int = Field(default=20, ge=1)
    header: str = ""
    width_px: int | None = None
    #: Presentation-only: a hidden column is still evaluated (later columns
    #: may reference it via ColumnRef) but is omitted from the grid and the
    #: xlsx export. Never feed this into evaluation — dropping the column
    #: would shift ColumnRef indices and the expand-slot arithmetic.
    hidden: bool = False
    #: JSON-export settings; `None` means "all defaults", which keeps saved
    #: payloads clean for the overwhelming majority of columns.
    json_export: JsonColumnOptions | None = None
    #: Export overrides (inclusion, xlsx header). `None` means "all defaults",
    #: which keeps saved payloads clean for the overwhelming majority.
    export: ColumnExportOptions | None = None


class ScriptColumn(BaseModel):
    kind: Literal["script"] = "script"
    source: ColumnSource = Field(default_factory=RowSlot)
    snippet: SnippetSource = Field(default_factory=SnippetSource)
    mode: Literal["collapse", "expand"] = "collapse"
    keep_empty: bool = True
    header: str = ""
    width_px: int | None = None
    #: Presentation-only: a hidden column is still evaluated (later columns
    #: may reference it via ColumnRef) but is omitted from the grid and the
    #: xlsx export. Never feed this into evaluation — dropping the column
    #: would shift ColumnRef indices and the expand-slot arithmetic.
    hidden: bool = False
    #: JSON-export settings; `None` means "all defaults", which keeps saved
    #: payloads clean for the overwhelming majority of columns.
    json_export: JsonColumnOptions | None = None
    #: Export overrides (inclusion, xlsx header). `None` means "all defaults",
    #: which keeps saved payloads clean for the overwhelming majority.
    export: ColumnExportOptions | None = None


Column = Annotated[
    ElementColumn | PropertyColumn | NavigationColumn | ScriptColumn,
    Field(discriminator="kind"),
]


class TableDefinition(BaseModel):
    schema_version: int = SCHEMA_VERSION
    row_source: RowSource
    columns: list[Column] = Field(min_length=1, max_length=MAX_COLUMNS)
    default_cell_mode: Literal["collapse", "expand"] = "collapse"
    #: Presentation flag: render a 1-based "#" first column in the grid and
    #: prepend the same column to the xlsx export. Not a real column — it
    #: never participates in ColumnRef indexing, sorting, or evaluation.
    show_row_numbers: bool = False
    #: Output order for the export, as definition column indices, with
    #: `export_layout.ROW_NUMBER_SLOT` (-1) standing for the row-number
    #: pseudo-column. `[]` = definition order. NOT an evaluation order:
    #: column order is structural (backward-only `ColumnRef`, positional
    #: expand slots) and is never permuted — this reorders the OUTPUT.
    #: Normalized rather than validated (see `export_layout.normalized_order`):
    #: a stale list left behind by a column insert or remove must degrade to a
    #: sensible order, not 422 the whole export.
    export_order: list[int] = Field(default_factory=list)
    export_row_number: RowNumberExportOptions | None = None
    #: JSON-export split settings; `None` = single-document export (today's
    #: behavior, and the no-migration guarantee for existing payloads).
    json_split: JsonSplitOptions | None = None

    @model_validator(mode="after")
    def _validate_sources(self) -> TableDefinition:
        is_chains = self.row_source.kind == "chains"
        for i, col in enumerate(self.columns):
            src = col.source
            # backward-only column refs
            if isinstance(src, ColumnRef) and src.index >= i:
                raise ValueError(
                    f"column {i} sources column {src.index} (must be < {i})"
                )
            if isinstance(src, ColumnRef) and src.step_index is not None:
                if self.columns[src.index].kind != "navigation":
                    raise ValueError(
                        f"column {i}: source step_index requires the referenced "
                        "column to be a navigation column"
                    )
            # chain_index only on a chains row source
            if isinstance(src, RowSlot) and src.chain_index != 0 and not is_chains:
                raise ValueError("chain_index != 0 requires a chains row source")
            # is the source element-producing, and is it single-binding?
            producing, single = self._source_arity(src)
            if col.kind == "navigation" and not producing:
                raise ValueError(
                    f"column {i}: navigation source is not element-producing"
                )
            if col.kind == "element" and not producing:
                raise ValueError(
                    f"column {i}: element column needs an element-producing source"
                )
            if col.kind == "element" and not single:
                raise ValueError(
                    f"column {i}: element column needs a single-binding source"
                )
            if col.kind == "property" and col.mode == "expand" and not single:
                raise ValueError(
                    f"column {i}: expanded property needs a single-binding source"
                )
        return self

    def _source_arity(self, src: ColumnSource) -> tuple[bool, bool]:
        """(element_producing, single_binding) for a column source.

        A row slot is always element-producing and single. A ColumnRef inherits
        from the referenced column: element/expand columns are single-binding
        elements; a collapse navigation column is multi-binding elements; a
        property column is not element-producing.
        """
        if isinstance(src, RowSlot):
            return True, True
        ref = self.columns[src.index]
        if ref.kind == "element":
            return True, True
        if ref.kind == "navigation":
            # a step-index override re-projects the chains and can return MANY
            # elements per row even off an expand column — no longer single.
            single = ref.mode == "expand" and src.step_index is None
            return True, single
        if ref.kind == "script":
            # Element-capable at RUNTIME: value() may return Element(s); a
            # scalar result simply binds nothing downstream (tolerant — the
            # return type is not statically knowable). Like a navigation
            # column, collapse is multi-binding, expand promotes one binding
            # per row.
            return True, ref.mode == "expand"
        # property column
        return False, ref.mode == "expand"


TABLE_ADAPTER: TypeAdapter[TableDefinition] = TypeAdapter(TableDefinition)

File: engine/src/table/schema.ts
/**
 * The table definitions of `core/table/schema.py`. A row is a tuple of
 * bindings: the row source contributes one slot (scope, navigation) or one
 * per chain step (chains), and every expand column one more. A column maps
 * over its source — a row slot or an earlier column — and keeps the values
 * in one cell (collapse) or promotes each to a slot of its own (expand).
 */
import { ReadError } from '../read/errors.ts';
import { readCriteria, type Criterion } from '../search/criteria.ts';
import { pyRepr } from '../value/repr.ts';
import { PyFloat } from '../value/types.ts';
import {
	readNavigation,
	type NavigationDefinition,
	type SnippetSource
} from '../navigation/schema.ts';

/** The most columns a table may hold. */
export const MAX_COLUMNS = 50;

/** A saved navigation (`ref`) or an inline one; neither is an unconfigured source, reaching nothing. */
export type NavigationSource = { ref: string | null; definition: NavigationDefinition | null };

export type ScopeRows = { kind: 'scope'; types: string[]; criteria: Criterion[] };

/** The elements at `step_index` of the navigation's chains (none: the last). */
export type NavigationRows = {
	kind: 'navigation';
	navigation: NavigationSource;
	step_index: number | null;
};

/** Whole chains, one slot per step; `unique` keeps the first chain to each terminal. */
export type ChainRows = { kind: 'chains'; navigation: NavigationSource; unique: boolean };

export type RowSource = ScopeRows | NavigationRows | ChainRows;

export type RowSlot = { kind: 'row'; chain_index: number };

/** An earlier column; `step_index` re-projects a navigation column's chains at that step. */
export type ColumnRef = { kind: 'column'; index: number; step_index: number | null };

export type ColumnSource = RowSlot | ColumnRef;

export type CellMode = 'collapse' | 'expand';

/** A named earlier column a script column reads for the same row. */
export type ScriptInput = { name: string; ref: ColumnRef };

export type JsonColumnOptions = {
	key: string;
	item_key: string;
	value: 'name' | 'id' | 'object';
	group: boolean;
	single: boolean;
};

export type ColumnExportOptions = { include: boolean | null; header: string };

/** What a column carries for display and export only. */
type Presentation = {
	header: string;
	width_px: number | null;
	hidden: boolean;
	json_export: JsonColumnOptions | null;
	export: ColumnExportOptions | null;
};

export type ElementColumn = { kind: 'element'; source: ColumnSource } & Presentation;

export type PropertyColumn = {
	kind: 'property';
	source: ColumnSource;
	name: string;
	mode: CellMode;
	keep_empty: boolean;
} & Presentation;

export type NavigationColumn = {
	kind: 'navigation';
	source: ColumnSource;
	navigation: NavigationSource;
	step_index: number | null;
	mode: CellMode;
	keep_empty: boolean;
	sort_mode: 'value' | 'count';
	cell_cap: number;
} & Presentation;

/** Its snippet is not read further: a table that reaches a script runs on the server. */
export type ScriptColumn = {
	kind: 'script';
	source: ColumnSource;
	snippet: SnippetSource;
	inputs: ScriptInput[];
	mode: CellMode;
	keep_empty: boolean;
} & Presentation;

export type Column = ElementColumn | PropertyColumn | NavigationColumn | ScriptColumn;

export type SortKey = { column: number; direction: 'asc' | 'desc' };

export type RowNumberExportOptions = { include: boolean; header: string; key: string };

export type JsonSplitOptions = { enabled: boolean; filename_template: string };

export type TableDefinition = {
	schema_version: number | bigint;
	row_source: RowSource;
	columns: Column[];
	default_cell_mode: CellMode;
	show_row_numbers: boolean;
	export_order: number[];
	display_order: number[];
	sort: SortKey[];
	export_row_number: RowNumberExportOptions | null;
	json_split: JsonSplitOptions | null;
	transform: SnippetSource | null;
};

// -- reading -------------------------------------------------------------------

type Doc = { readonly [key: string]: unknown };

const isDoc = (value: unknown): value is Doc =>
	typeof value === 'object' &&
	value !== null &&
	!Array.isArray(value) &&
	!(value instanceof PyFloat);

function refuse(where: string, message: string): never {
	throw new ReadError(422, `${where}: ${message}`);
}

// An absent key takes the default; a `null` does not, as pydantic has it.
const field = (d: Doc, key: string, fallback?: unknown): unknown =>
	Object.hasOwn(d, key) && d[key] !== undefined ? d[key] : fallback;

function doc(raw: unknown, where: string): Doc {
	if (!isDoc(raw)) refuse(where, 'must be an object');
	return raw;
}

const isInt = (value: unknown): value is number =>
	typeof value === 'number' && Number.isInteger(value);

function str(d: Doc, key: string, where: string, fallback?: string): string {
	const value = field(d, key, fallback);
	if (typeof value !== 'string') refuse(`${where}.${key}`, 'must be a string');
	return value;
}

function optionalStr(d: Doc, key: string, where: string): string | null {
	const value = field(d, key, null);
	if (value !== null && typeof value !== 'string') {
		refuse(`${where}.${key}`, 'must be a string or null');
	}
	return value;
}

function bool(d: Doc, key: string, where: string, fallback: boolean): boolean {
	const value = field(d, key, fallback);
	if (typeof value !== 'boolean') refuse(`${where}.${key}`, 'must be a boolean');
	return value;
}

function int(d: Doc, key: string, where: string, min: number, fallback?: number): number {
	const value = field(d, key, fallback);
	if (!isInt(value) || value < min) {
		refuse(`${where}.${key}`, `must be an integer of at least ${min}`);
	}
	return value;
}

function optionalInt(d: Doc, key: string, where: string): number | null {
	const value = field(d, key, null);
	if (value !== null && !isInt(value)) refuse(`${where}.${key}`, 'must be an integer or null');
	return value;
}

function oneOf<T extends string>(
	d: Doc,
	key: string,
	where: string,
	options: readonly T[],
	fallback?: T
): T {
	const value = field(d, key, fallback);
	if (!options.includes(value as T)) {
		refuse(`${where}.${key}`, `must be one of ${options.join(', ')}`);
	}
	return value as T;
}

function names(d: Doc, key: string, where: string): string[] {
	const value = field(d, key, []);
	if (!Array.isArray(value) || !value.every((name) => typeof name === 'string')) {
		refuse(`${where}.${key}`, 'must be a list of strings');
	}
	return [...(value as string[])];
}

function ints(d: Doc, key: string, where: string): number[] {
	const value = field(d, key, []);
	if (!Array.isArray(value) || !value.every(isInt)) {
		refuse(`${where}.${key}`, 'must be a list of integers');
	}
	return [...(value as number[])];
}

function list(d: Doc, key: string, where: string): readonly unknown[] {
	const value = field(d, key, []);
	if (!Array.isArray(value)) refuse(`${where}.${key}`, 'must be a list');
	return value;
}

/** An optional sub-document: `null` when absent or null, else read by `read`. */
function optionalDoc<T>(
	d: Doc,
	key: string,
	where: string,
	read: (sub: Doc, at: string) => T
): T | null {
	const value = field(d, key, null);
	if (value === null) return null;
	const at = `${where}.${key}`;
	return read(doc(value, at), at);
}

function readNavigationSource(raw: unknown, where: string): NavigationSource {
	const d = doc(raw, where);
	const ref = optionalStr(d, 'ref', where);
	const inline = field(d, 'definition', null);
	const definition = inline === null ? null : readNavigation(inline, `${where}.definition`);
	if (ref !== null && definition !== null) {
		refuse(where, 'provide at most one of `ref` / `definition`');
	}
	return { ref, definition };
}

function readSnippetSource(d: Doc, where: string): SnippetSource {
	const ref = optionalStr(d, 'ref', where);
	const definition = field(d, 'definition', null);
	if (definition !== null && !isDoc(definition)) {
		refuse(`${where}.definition`, 'must be an object or null');
	}
	if (ref !== null && definition !== null) {
		refuse(where, 'provide at most one of `ref` / `definition`');
	}
	return { ref, definition };
}

function readRowSource(raw: unknown, where: string): RowSource {
	const d = doc(raw, where);
	const kind = oneOf(d, 'kind', where, ['scope', 'navigation', 'chains'] as const);
	if (kind === 'scope') {
		return {
			kind,
			types: names(d, 'types', where),
			criteria: readCriteria(field(d, 'criteria', []), `${where}.criteria`)
		};
	}
	const navigation = readNavigationSource(field(d, 'navigation'), `${where}.navigation`);
	if (kind === 'navigation') {
		return { kind, navigation, step_index: optionalInt(d, 'step_index', where) };
	}
	return { kind, navigation, unique: bool(d, 'unique', where, false) };
}

function readColumnRef(d: Doc, where: string): ColumnRef {
	oneOf(d, 'kind', where, ['column'] as const, 'column');
	return {
		kind: 'column',
		index: int(d, 'index', where, 0),
		step_index: optionalInt(d, 'step_index', where)
	};
}

function readSource(d: Doc, where: string): ColumnSource {
	const raw = field(d, 'source');
	const at = `${where}.source`;
	if (raw === undefined) return { kind: 'row', chain_index: 0 };
	const s = doc(raw, at);
	const kind = oneOf(s, 'kind', at, ['row', 'column'] as const);
	if (kind === 'row') return { kind, chain_index: int(s, 'chain_index', at, 0, 0) };
	return readColumnRef(s, at);
}

// Python's `keyword.kwlist`.
const KEYWORDS = new Set(
	(
		'False None True and as assert async await break class continue def del elif else except ' +
		'finally for from global if import in is lambda nonlocal not or pass raise return try while ' +
		'with yield'
	).split(' ')
);

const IDENTIFIER = /^[\p{XID_Start}_]\p{XID_Continue}*$/u;

function readInput(raw: unknown, where: string): ScriptInput {
	const d = doc(raw, where);
	const name = str(d, 'name', where);
	if (!IDENTIFIER.test(name) || KEYWORDS.has(name)) {
		refuse(`${where}.name`, `input name ${pyRepr(name)} is not a valid identifier`);
	}
	const at = `${where}.ref`;
	return { name, ref: readColumnRef(doc(field(d, 'ref'), at), at) };
}

function readJsonOptions(d: Doc, where: string): JsonColumnOptions {
	return {
		key: str(d, 'key', where, ''),
		item_key: str(d, 'item_key', where, ''),
		value: oneOf(d, 'value', where, ['name', 'id', 'object'] as const, 'name'),
		group: bool(d, 'group', where, false),
		single: bool(d, 'single', where, false)
	};
}

function readExportOptions(d: Doc, where: string): ColumnExportOptions {
	const include = field(d, 'include', null);
	if (include !== null && typeof include !== 'boolean') {
		refuse(`${where}.include`, 'must be a boolean or null');
	}
	return { include, header: str(d, 'header', where, '') };
}

function readPresentation(d: Doc, where: string): Presentation {
	return {
		header: str(d, 'header', where, ''),
		width_px: optionalInt(d, 'width_px', where),
		hidden: bool(d, 'hidden', where, false),
		json_export: optionalDoc(d, 'json_export', where, readJsonOptions),
		export: optionalDoc(d, 'export', where, readExportOptions)
	};
}

const MODES = ['collapse', 'expand'] as const;

function readColumn(raw: unknown, where: string): Column {
	const d = doc(raw, where);
	const kind = oneOf(d, 'kind', where, ['element', 'property', 'navigation', 'script'] as const);
	const source = readSource(d, where);
	switch (kind) {
		case 'element':
			return { kind, source, ...readPresentation(d, where) };
		case 'property':
			return {
				kind,
				source,
				name: str(d, 'name', where),
				mode: oneOf(d, 'mode', where, MODES, 'collapse'),
				keep_empty: bool(d, 'keep_empty', where, true),
				...readPresentation(d, where)
			};
		case 'navigation':
			return {
				kind,
				source,
				navigation: readNavigationSource(field(d, 'navigation'), `${where}.navigation`),
				step_index: optionalInt(d, 'step_index', where),
				mode: oneOf(d, 'mode', where, MODES, 'collapse'),
				keep_empty: bool(d, 'keep_empty', where, true),
				sort_mode: oneOf(d, 'sort_mode', where, ['value', 'count'] as const, 'value'),
				cell_cap: int(d, 'cell_cap', where, 1, 20),
				...readPresentation(d, where)
			};
		case 'script':
			return {
				kind,
				source,
				snippet: readSnippetSource(
					doc(field(d, 'snippet', {}), `${where}.snippet`),
					`${where}.snippet`
				),
				inputs: list(d, 'inputs', where).map((input, i) =>
					readInput(input, `${where}.inputs[${i}]`)
				),
				mode: oneOf(d, 'mode', where, MODES, 'collapse'),
				keep_empty: bool(d, 'keep_empty', where, true),
				...readPresentation(d, where)
			};
	}
}

function readSortKey(raw: unknown, where: string): SortKey {
	const d = doc(raw, where);
	return {
		column: int(d, 'column', where, 0),
		direction: oneOf(d, 'direction', where, ['asc', 'desc'] as const, 'asc')
	};
}

// -- static checks ---------------------------------------------------------------

/** Whether a source yields elements, and whether one per row. */
function sourceArity(columns: readonly Column[], source: ColumnSource): [boolean, boolean] {
	if (source.kind === 'row') return [true, true];
	const ref = columns[source.index]!;
	if (ref.kind === 'element') return [true, true];
	// A step index re-projects the chains: many elements a row, even off an expand column.
	if (ref.kind === 'navigation') return [true, ref.mode === 'expand' && source.step_index === null];
	return [true, ref.mode === 'expand'];
}

function checkInputs(
	columns: readonly Column[],
	i: number,
	col: ScriptColumn,
	where: string
): void {
	const seen = new Set<string>();
	col.inputs.forEach((input, k) => {
		if (seen.has(input.name)) {
			refuse(`${where}.inputs`, `column ${i}: duplicate input name ${pyRepr(input.name)}`);
		}
		seen.add(input.name);
		const at = `${where}.inputs[${k}].ref`;
		if (input.ref.index >= i) {
			refuse(
				at,
				`column ${i}: input ${pyRepr(input.name)} references column ${input.ref.index} (must be < ${i})`
			);
		}
		if (input.ref.step_index !== null && columns[input.ref.index]!.kind !== 'navigation') {
			refuse(
				at,
				`column ${i}: input ${pyRepr(input.name)} step_index requires the referenced column to be a navigation column`
			);
		}
	});
}

/**
 * The core's `_validate_sources`, in its order: a column ref points strictly
 * back; a ref's step index needs a navigation column; a row slot past 0 needs
 * chains; an element column needs one element a row. The `value()` arity of an
 * inline snippet is not checked: such a table reaches a script and is read
 * from the server, which checks it.
 */
function checkSources(defn: TableDefinition, where: string): void {
	const chains = defn.row_source.kind === 'chains';
	const { columns } = defn;
	columns.forEach((col, i) => {
		const at = `${where}.columns[${i}]`;
		const { source } = col;
		if (source.kind === 'column' && source.index >= i) {
			refuse(`${at}.source`, `column ${i} sources column ${source.index} (must be < ${i})`);
		}
		if (source.kind === 'column' && source.step_index !== null) {
			if (columns[source.index]!.kind !== 'navigation') {
				refuse(
					`${at}.source`,
					`column ${i}: source step_index requires the referenced column to be a navigation column`
				);
			}
		}
		if (source.kind === 'row' && source.chain_index !== 0 && !chains) {
			refuse(`${at}.source`, 'chain_index != 0 requires a chains row source');
		}
		const [producing, single] = sourceArity(columns, source);
		if (col.kind === 'navigation' && !producing) {
			refuse(`${at}.source`, `column ${i}: navigation source is not element-producing`);
		}
		if (col.kind === 'element' && !producing) {
			refuse(`${at}.source`, `column ${i}: element column needs an element-producing source`);
		}
		if (col.kind === 'element' && !single) {
			refuse(`${at}.source`, `column ${i}: element column needs a single-binding source`);
		}
		if (col.kind === 'script') checkInputs(columns, i, col, at);
	});
}

/**
 * A table definition as a client sends it or an artifact holds it, read as
 * pydantic reads it in canonical JSON: every `kind` required, the defaults
 * filled, unknown keys ignored, then the core's static checks. Navigations are
 * read by `readNavigation`; snippets are kept, not read. What pydantic would
 * coerce is refused, in the engine's words where the core has none.
 */
export function readTableDefinition(raw: unknown, where: string): TableDefinition {
	const d = doc(raw, where);
	const schemaVersion = field(d, 'schema_version', 1);
	if (!(isInt(schemaVersion) || typeof schemaVersion === 'bigint')) {
		refuse(`${where}.schema_version`, 'must be an integer');
	}
	const rowSource = readRowSource(field(d, 'row_source'), `${where}.row_source`);
	const rawColumns = field(d, 'columns');
	if (!Array.isArray(rawColumns)) refuse(`${where}.columns`, 'must be a list');
	if (rawColumns.length < 1 || rawColumns.length > MAX_COLUMNS) {
		refuse(`${where}.columns`, `must hold 1 to ${MAX_COLUMNS} columns`);
	}
	const defn: TableDefinition = {
		schema_version: schemaVersion,
		row_source: rowSource,
		columns: rawColumns.map((col: unknown, i) => readColumn(col, `${where}.columns[${i}]`)),
		default_cell_mode: oneOf(d, 'default_cell_mode', where, MODES, 'collapse'),
		show_row_numbers: bool(d, 'show_row_numbers', where, false),
		export_order: ints(d, 'export_order', where),
		display_order: ints(d, 'display_order', where),
		sort: list(d, 'sort', where).map((key, i) => readSortKey(key, `${where}.sort[${i}]`)),
		export_row_number: optionalDoc(d, 'export_row_number', where, (sub, at) => ({
			include: bool(sub, 'include', at, true),
			header: str(sub, 'header', at, ''),
			key: str(sub, 'key', at, '')
		})),
		json_split: optionalDoc(d, 'json_split', where, (sub, at) => ({
			enabled: bool(sub, 'enabled', at, false),
			filename_template: str(sub, 'filename_template', at, '')
		})),
		transform: optionalDoc(d, 'transform', where, readSnippetSource)
	};
	checkSources(defn, where);
	return defn;
}
